Workaround for DAO having garbled content after RPL forwarding error

arurke opened the pull request:
This PR
1. Proposes a workaround to RPL issue #2377 
2. ~~Adds a test that showcases #2377~~ The test for the issue was added separately in #2548. This PR simply reverts the logic in the test such that it passes only if the issue is resolved.

In short, #2377 describes how No-Path DAOs that are sent when detecting RPL forwarding error will have garbled content and the receiver will discard it with error-log: `icmpv6 bad checksum`. The workaround adds a small delay (next tick) to the generation of the DAO, and this resolves the issue. I am no uIP/RPL expert and I have not identified the root cause as to why the content is garbled (see some ideas in #2377). As such, I will have no trouble if we don't merge this and rather leave it for others to work onward from.

For further details see test-description in #2548 and issue #2377.

pjonsson commented on tests/15-rpl-classic/code-simple-tx-rx/simple-node.c:
```diff
@@ -0,0 +1,112 @@
+{
+  static uint32_t rx_count = 0;
+  rx_count++;
+  LOG_INFO("Data %u received on port %d from port %d with length %d\n",
```
Use the `PRI` family of defines from `inttypes.h` to print fixed size types. (`PRIu32`/`PRIu16` for these).

arurke replied:
Fixed in #2548

pjonsson commented on tests/15-rpl-classic/js/13-rpl-dao-garbled.js:
```diff
@@ -0,0 +1,41 @@
+      if(forwarding_error_occurred && no_path_dao_received) {
+        log.testOK();
+      }
+      else {
```
`log.testOK()` will terminate, so the else-block can be removed.

arurke replied:
Fixed in #2548

pjonsson commented on tests/15-rpl-classic/js/12-rpl-rank-error.js:
```diff
@@ -0,0 +1,15 @@
+TIMEOUT(600000, log.testFailed());
```
I don't think the second parameter is required, a timed out test should fail.

arurke replied:
Fixed in #2548

pjonsson commented on tests/15-rpl-classic/js/12-rpl-rank-error.js:
```diff
@@ -0,0 +1,15 @@
+    log.testFailed();
+  }
+
+  else if(msg.contains("Received all packets")) {
```
Previous block returns, so the `else` can be removed.

arurke replied:
Fixed in #2548

pjonsson commented on os/net/routing/rpl-classic/rpl.h:
```diff
@@ -262,6 +262,10 @@ struct rpl_instance {
+  struct ctimer unicast_dao_timer;
+  rpl_parent_t *unicast_dao_target;
+  uip_ipaddr_t *unicast_dao_prefix;
+  uint8_t unicast_dao_lifetime;
```
I'm guessing there will be padding between `unicast_dao_lifetime` and `dao_retransmit_timer`. Putting these fields after `dao_retransmit_timer` should avoid that.

arurke replied:
:+1:

pjonsson commented:
The test case is valuable to get into develop for tracking the behavior, so we'll know if this issue is accidentally fixed by something else.

Can you split out the test cases to a separate PR that can be merged? Just invert the logic for passing/failing testcase, and put a FIXME in a comment above that.

arurke commented:
> Can you split out the test cases to a separate PR that can be merged? Just invert the logic for passing/failing testcase, and put a FIXME in a comment above that.

🤦‍♂️ This is a great idea @pjonsson. Added https://github.com/contiki-ng/contiki-ng/pull/2548

simonduq commented:
Hi!

This was my mistake, very sorry about it, re-opening this PR now.

What happened is the following: I was switching the base branch for this repository https://github.com/wittra/contiki-ng from develop to wittra. But accidentally, I made the change on the wrong repo (this repo) and instead of switching the base branch I renamed it. And somehow github deleted develop and closed all PRs...

Many apologies for this mishap 🙏; I haven't contributed in a while.. but now at least everybody got some notification from me :p

arurke commented:
I have now rebased and removed the tests that were added via https://github.com/contiki-ng/contiki-ng/pull/2548.

As mentioned I believe there is more work to be done on the issue (ref. my explanation above of why calling this a workaround) - at the same time I believe this PR is conservative and is unlikely to do any harm.

pjonsson commented:
@arurke can you please rebase this to latest develop?

arurke commented:
👍 rebased
